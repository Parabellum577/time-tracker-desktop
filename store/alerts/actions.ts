import { createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { IAlert } from './types'

export const showAlertMessage = createAsyncAction(
  'notifications/SHOW_ALERT_MESSAGE_REQUEST',
  'notifications/SHOW_ALERT_MESSAGE_SUCCESS',
  'notifications/SHOW_ALERT_MESSAGE_FAILURE',
)<IAlert, IAlert, IError>()

export const clearAlertMessage = createAsyncAction(
  'notifications/CLEAR_ALERT_MESSAGE_REQUEST',
  'notifications/CLEAR_ALERT_MESSAGE_SUCCESS',
  'notifications/CLEAR_ALERT_MESSAGE_FAILURE',
)<void, void, IError>()
