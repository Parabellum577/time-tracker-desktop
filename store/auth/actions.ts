import { createAction, createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'

export const loginUser = createAsyncAction('auth/LOGIN_REQUEST', 'auth/LOGIN_SUCCESS', 'auth/LOGIN_FAILURE')<
  ILoginUserPayload,
  void,
  IError
>()

export const loginUserViaOAuth = createAsyncAction(
  'auth/LOGIN_VIA_OAUTH_REQUEST',
  'auth/LOGIN_VIA_OAUTH_SUCCESS',
  'auth/LOGIN_VIA_OAUTH_FAILURE',
)<void, void, IError>()

export const otsLogin = createAsyncAction('auth/OTS_LOGIN_REQUEST', 'auth/OTS_LOGIN_SUCCESS', 'auth/OTS_LOGIN_FAILURE')<
  ILoginOTSPayload,
  ILoginOTSPayload,
  IError
>()

export const loadingDataAfterLogin = createAction('auth/LOADING_DATA_AFTER_LOGIN', resolve => (isLoading: boolean) =>
  resolve(isLoading),
)

export const signOut = createAction('auth/SIGN_OUT', resolve => () => resolve())
