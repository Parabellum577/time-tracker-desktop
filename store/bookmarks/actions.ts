import { createAsyncAction, createAction } from 'typesafe-actions'
import { IError, IFullIssue } from '@types'

export const getBookmarks = createAsyncAction(
  'bookmarks/GET_BOOKMARKS_REQUEST',
  'bookmarks/GET_BOOKMARKS_SUCCESS',
  'bookmarks/GET_BOOKMARKS_FAILURE',
)<void, IFullIssue[], IError>()

export const handleBookmark = createAsyncAction(
  'bookmarks/HANDLE_BOOKMARK_REQUEST',
  'bookmarks/HANDLE_BOOKMARK_SUCCESS',
  'bookmarks/HANDLE_BOOKMARK_FAILURE',
)<IFullIssue, IFullIssue, IError>()

export const clearBookmarks = createAction('plannings/CLEAR_BOOKMARKS', action => () => action())
