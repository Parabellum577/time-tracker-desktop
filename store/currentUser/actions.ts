import { createAction, createAsyncAction } from 'typesafe-actions'
import { IUser, IError, IUpdateUser, ISettingsItem } from '@types'

export const getCurrentUser = createAsyncAction(
  'user/GET_CURRENT_USER_REQUEST',
  'user/GET_CURRENT_USER_SUCCESS',
  'user/GET_CURRENT_USER_FAILURE',
)<void, IUser, IError>()

export const updateCurrentUser = createAction('user/UPDATE_CURRENT_USER', action => (userData: IUpdateUser) =>
  action({
    Avatar: userData.Avatar,
    FirstName: userData.FirstName,
    LastName: userData.LastName,
    Phone: userData.Phone,
    Skype: userData.Skype,
    Role: userData.Role,
  }),
)

export const handleUserSettings = createAsyncAction(
  'user/HANDLE_USER_SETTINGS_REQUEST',
  'user/HANDLE_USER_SETTINGS_SUCCESS',
  'user/HANDLE_USER_SETTINGS_FAILURE',
)<ISettingsItem[], ISettingsItem[], IError>()

export const getUserSettings = createAsyncAction(
  'user/GET_USER_SETTINGS_REQUEST',
  'user/GET_USER_SETTINGS_SUCCESS',
  'user/GET_USER_SETTINGS_FAILURE',
)<void, void, IError>()

export const setDefaultUserSettings = createAction('user/SET_DEFAULT_USER_SETTINGS', action => action)

export const activateUserAccount = createAction('user/ACTIVATE_USER_ACCOUNT', action => action)
