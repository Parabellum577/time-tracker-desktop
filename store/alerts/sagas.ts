import * as _ from 'lodash'
import { SagaIterator } from 'redux-saga'
import { put, take, select, spawn, delay } from 'redux-saga/effects'
import { getType } from 'typesafe-actions'

import { IRootState } from '@store/rootReducer'
import * as actions from './actions'
import { IAlert } from './types'

export const alertsArrayFromStore = (state: IRootState) => state.alerts.alertsArray

export function* showAlertsSaga(): SagaIterator {
  while (true) {
    const { payload } = yield take(getType(actions.showAlertMessage.request))
    let alertsArray: IAlert[] = yield select(alertsArrayFromStore)
    if (_.isEmpty(alertsArray)) {
      yield spawn(function*() {
        while (true) {
          yield delay(3000)
          alertsArray = yield select(alertsArrayFromStore)
          if (_.isEmpty(alertsArray)) {
            return
          }
          yield put(actions.clearAlertMessage.success())
        }
      })
    }
    yield put(actions.showAlertMessage.success(payload))
  }
}

export function* handleShowAlerts(alert: IAlert) {
  yield put(actions.showAlertMessage.success(alert))
}

export function* clearAlertsSaga(): SagaIterator {
  while (true) {
    yield take(getType(actions.clearAlertMessage.request))
    yield put(actions.clearAlertMessage.success())
  }
}
