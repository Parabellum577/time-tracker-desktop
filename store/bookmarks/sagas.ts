import { call, take, race, put, select } from 'redux-saga/effects'
import { SagaIterator } from 'redux-saga'
import { getType } from 'typesafe-actions'

import * as loginActions from '../auth/actions'
import * as alertActions from '../alerts/actions'
import * as actions from './actions'
import api from '@api'
import { IBasicIssue, IFullIssue } from '@types'
import extendIssues from '@services/extendIssues'
import { IRootState } from '@store/rootReducer'

export function* getBookmarksSaga(): SagaIterator {
  while (true) {
    yield race({
      getBookmarksAction: take(getType(actions.getBookmarks.request)),
      loginUserAction: take(getType(loginActions.loginUser.success)),
    })
    try {
      const basicBookmarks: IBasicIssue[] = yield call(api.tracker.GetBookmarks)
      const bookmarks = yield call(extendIssues, basicBookmarks)
      yield put(actions.getBookmarks.success(bookmarks))
    } catch (error) {
      console.error(error)
    }
  }
}

export const bookmarksArrayFromStore = (state: IRootState) => state.bookmarks.extendedBookmarks

export function* handleBookmarkSaga(): SagaIterator {
  while (true) {
    const { payload } = yield take(getType(actions.handleBookmark.request))
    try {
      const bookmarks: IFullIssue[] = yield select(bookmarksArrayFromStore)
      const isRemove = bookmarks.some(bookmark => bookmark.ID === payload.ID)
      const bookmarkPayload = {
        TrackerID: payload.TrackerID,
        ProjectID: payload.ProjectID,
        IssueID: payload.ID,
      }
      yield put(actions.handleBookmark.success(payload))
      if (isRemove) {
        yield call(api.tracker.RemoveBookmark, bookmarkPayload)
        yield put(
          alertActions.showAlertMessage.request({
            alertType: 'success',
            alertMessage: 'Removed from Favorites',
          }),
        )
      } else {
        yield call(api.tracker.CreateBookmark, bookmarkPayload)
        yield put(
          alertActions.showAlertMessage.request({
            alertType: 'success',
            alertMessage: 'Added to Favorites',
          }),
        )
      }
    } catch (error) {
      console.error('handleBookmarkSaga: ', error)
    }
  }
}
