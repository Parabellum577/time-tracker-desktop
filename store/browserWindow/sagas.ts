import { ipcMain, app } from 'electron'
import { call, put, race, take, delay } from 'redux-saga/effects'
import { getActivePlanning, getSyncState } from '@services/helpers'
import { IPlanning } from '@services/types'
import * as planningsActions from '@store/plannings/actions'
import { getType } from 'typesafe-actions'
import persistStore from '@store/persistStore'
import { REASON_FOR_QUIT_KEY } from '@electronMain/src/helpers/memoryLeakDetection'
import { syncSpentSlices } from '@store/timeTracker/sagas'
import { getWindow } from '@electronMain/src/helpers/browserWindow'

export function* stopPlanningBeforeCloseAppliction() {
  yield call(
    () =>
      new Promise(res => {
        ipcMain.once('close-application', res)
      }),
  )
  const win = getWindow()
  if (win) {
    win.close()
  }
  console.log('CLOSE_APP_HANDLER')
  const activePlanning: IPlanning = yield call(getActivePlanning)
  const isMaster: IPlanning = yield call(getSyncState)

  if (activePlanning && activePlanning.ID && isMaster) {
    yield put(planningsActions.stopPlanning.request())
    yield race([
      take(getType(planningsActions.stopPlanning.success)),
      take(getType(planningsActions.stopPlanning.failure)),
    ])
  }
  try {
    yield race([call(syncSpentSlices, true), delay(30 * 1000)])
  } catch (error) {
    console.log('function*stopPlanningBeforeCloseAppliction -> error', error)
  }
  console.log('quit')
  persistStore.set(REASON_FOR_QUIT_KEY, 'CLOSE_APP')
  app.quit()
}
