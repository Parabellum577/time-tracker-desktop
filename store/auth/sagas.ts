import { SagaIterator } from 'redux-saga'
import { all, call, fork, put, take, race, delay, select, cancel } from 'redux-saga/effects'
import { ActionType, getType } from 'typesafe-actions'
import { push } from 'connected-react-router'

import * as alertActions from '@store/alerts/actions'
import * as userActions from '@store/currentUser/actions'
import * as planningsActions from '@store/plannings/actions'
import * as bookmarksActions from '@store/bookmarks/actions'
import * as productivityActions from '@store/productivity/actions'
import * as projectsActions from '@store/projects/actions'
import * as recentActions from '@store/recents/actions'
import * as timelineActions from '@store/timeline/actions'
import * as screenshotActions from '@store/screenshot/actions'
import * as trackerActions from '@store/trackers/actions'
import * as timeTrackerActions from '@store/timeTracker/actions'
import * as synchronizationActions from '@store/synchronization/actions'
import * as actions from './actions'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'
import { setAuthData, getAuthData, IAuthData, updateRefreshToken } from './utils'
import api from '@api'
import { IRootState } from '@store/rootReducer'
import { IPlanning } from '@services/types'
import time from '@services/time'
import { getActivePlanning, getSyncState } from '@services/helpers'
import { createWindow, getWindow } from '@electronMain/src/helpers/browserWindow'
import { sendEvent } from '@electronMain/src/analytics/facade'
import { syncSpentSlices } from '@store/timeTracker/sagas'

const getAuthorizedFromStore = (state: IRootState) => state.auth.authorized
const isOnlineFromStore = (state: IRootState) => state.internetConnection.isOnline

export function* authSaga() {
  while (true) {
    let authData = getAuthData()
    const [credentialsAction, otsAction] = yield race([
      take(actions.loginUser.request), // Login by credentials
      take(actions.otsLogin.request), // Login by OTS
      take(getType(actions.loginUserViaOAuth.request)), // Login via OAuth
      call(() => (authData ? Promise.resolve() : new Promise(() => {}))), // Login by saved Refresh Token
    ])
    sendEvent('user', 'LOGIN_REQUEST')
    if (credentialsAction) {
      yield call(setAuthDataByCreds, credentialsAction.payload)
    } else if (otsAction) {
      yield call(setAuthDataByOTS, otsAction.payload)
    }

    authData = getAuthData()
    if (!authData) {
      yield put(push('/login'))
      continue
    }
    yield call(refreshAuth, authData.Refresh)
    if (!authData) {
      yield put(push('/login'))
      continue
    }
    yield put(userActions.getCurrentUser.request())
    const { error } = yield race({
      success: take(userActions.getCurrentUser.success),
      error: take(userActions.getCurrentUser.failure),
    })

    if (error) {
      yield put(
        alertActions.showAlertMessage.request({
          alertMessage: 'Oops! Something went wrong. Please relogin.',
          alertType: 'error',
        }),
      )
      continue
    }

    const taskDataFetchAfterLogin = yield fork(function*() {
      yield put(actions.loginUser.success())
      yield put(push('/main'))
      yield call(takeDataAfterLoading)
    })
    sendEvent('user', 'LOGIN_SUCCESS')
    yield race([call(refreshAuthToken), take(actions.signOut)])

    console.log('Logout')
    if (taskDataFetchAfterLogin) {
      cancel(taskDataFetchAfterLogin)
    }
    yield call(clearDataAfterLogout)
    yield put(push('/login'))
  }
}

function* setAuthDataByCreds({ Login, Password }: ILoginUserPayload): SagaIterator {
  try {
    const authData: IAuthData = yield call(api.user.LoginUser, {
      Login,
      Password,
    })
    yield call(setAuthData, authData)
    return authData
  } catch (error) {
    yield call(setAuthData, null)
    yield put(alertActions.showAlertMessage.request({ alertType: 'error', alertMessage: error.message }))
    yield put(actions.loginUser.failure(error))
    return null
  }
}

function* setAuthDataByOTS({ OTS }: ILoginOTSPayload): SagaIterator {
  const authorized = yield select(getAuthorizedFromStore)
  if (authorized) {
    return
  }

  try {
    const authData: IAuthData = yield call(api.user.OTSLogin, { OTS })
    yield call(setAuthData, authData)
    return authData
  } catch (error) {
    yield call(setAuthData, null)
    yield put(alertActions.showAlertMessage.request({ alertType: 'error', alertMessage: error.message }))
    yield put(actions.loginUser.failure(error))
    return null
  }
}

export function* takeDataAfterLoading(): SagaIterator {
  yield put(actions.loadingDataAfterLogin(true))
  yield fork(api.project.UpdateProjectsCache)
  yield all([
    put(planningsActions.getOpenPlannings.request()),
    put(planningsActions.getColdPlannings.request()),
    put(planningsActions.getClosedPlannings.request()),
    put(trackerActions.getUserTrackers.request()),
    put(timelineActions.getTimeline.request()),
    put(projectsActions.getProjects.request()),
    race([
      take(getType(planningsActions.getOpenPlannings.failure)),
      take(getType(planningsActions.getOpenPlannings.success)),
    ]),
    race([take(getType(projectsActions.getProjects.failure)), take(getType(projectsActions.getProjects.success))]),
    race([
      take(getType(planningsActions.getColdPlannings.failure)),
      take(getType(planningsActions.getColdPlannings.success)),
    ]),
    race([
      take(getType(planningsActions.getColdPlannings.failure)),
      take(getType(planningsActions.getColdPlannings.success)),
    ]),
    race([
      take(getType(planningsActions.getClosedPlannings.failure)),
      take(getType(planningsActions.getClosedPlannings.success)),
    ]),
    race([
      take(getType(trackerActions.getUserTrackers.failure)),
      take(getType(trackerActions.getUserTrackers.success)),
    ]),
    race([take(getType(timelineActions.getTimeline.failure)), take(getType(timelineActions.getTimeline.success))]),
    race([take(getType(userActions.getUserSettings.failure)), take(getType(userActions.getUserSettings.success))]),
  ])
  yield put(actions.loadingDataAfterLogin(false))
}

export function* refreshAuthToken(): SagaIterator {
  let authData = getAuthData()
  while (true) {
    if (!authData) {
      return
    }

    const isOnline: boolean = yield select(isOnlineFromStore)
    if (isOnline) {
      try {
        authData = yield call(refreshAuth, authData.Refresh)
        sendEvent('user', 'REFRESH_TOKEN')
      } catch (error) {
        console.error(error)
      }
    }
    const refreshAfter = authData ? (authData.Expires - time.now()) * 800 : 60000
    yield delay(refreshAfter > 0 ? refreshAfter : 60000)
  }
}

export function* refreshAuth(refresh: string): SagaIterator {
  try {
    const authData = yield call(updateRefreshToken, refresh)
    return authData
  } catch (e) {
    console.log('refreshAuth -> error:', e)
    console.log('refreshAuth -> error:', e.code, typeof e.code)
    const fatalCodes = [1, 102, 107, 206, 209, -32000]
    // INVALID_TOKEN
    // REFRESH_TOKEN_EXPIRED
    // USER_NOT_FOUND
    // MAIL_CONFIRMATION_EXPIRED
    // INVALID_REFRESH
    // crypto/rsa: verification error
    if (fatalCodes.includes(e.code)) {
      const message =
        e.code === 206
          ? 'Your account has expired... Please confirm your email to continue.'
          : 'Oops! Something went wrong. Please relogin.'
      yield put(
        alertActions.showAlertMessage.request({
          alertMessage: message,
          alertType: 'error',
        }),
      )
      const mainWindow = getWindow() || createWindow()
      mainWindow.show()
      mainWindow.focus()
      yield call(setAuthData, null)
      return null
    } else {
      return getAuthData()
    }
  }
}

export function* clearDataAfterLogout(): SagaIterator {
  const activePlanning: IPlanning = yield call(getActivePlanning)
  const isMaster: IPlanning = yield call(getSyncState)

  if (activePlanning && activePlanning.ID && isMaster) {
    yield put(planningsActions.stopPlanning.request())
    yield race([
      take(getType(planningsActions.stopPlanning.success)),
      take(getType(planningsActions.stopPlanning.failure)),
    ])
  }
  try {
    yield call(syncSpentSlices)
  } catch (e) {
    console.error('clearDataAfterLogout -> error', e)
  }
  yield put(planningsActions.clearPlannings())
  yield put(bookmarksActions.clearBookmarks())
  yield put(productivityActions.clearProductivities())
  yield put(projectsActions.clearProjects())
  yield put(recentActions.clearRecent())
  yield put(screenshotActions.deleteScreenshotStorages())
  yield put(trackerActions.clearUserTrackers())
  yield put(timeTrackerActions.removeAllSpentSlices())
  yield put(synchronizationActions.changeMasterStatus(true))
  yield call(setAuthData, null)
}
