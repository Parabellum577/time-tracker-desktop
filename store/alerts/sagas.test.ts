import { take, resetMocks, call, put } from '@test-helpers/effects.mock'
import { clearAlertsSaga, handleShowAlerts } from './sagas'
import * as actions from './actions'
import { SIMPLE_ALERT } from '@test-helpers/sample'

describe('[Store] Alert Sagas', () => {
  beforeEach(() => {
    resetMocks()
  })

  // it('[Success] Should handle showAlertMessage Request', () => {
  //   const newAlert = SIMPLE_ALERT
  //   const currentState = {
  //     alertsArray: [{ ...SIMPLE_ALERT }]
  //   }
  //   const generator = showAlertsSaga()
  //   generator.next()
  //   expect(take).toBeCalledTimes(1)
  //   expect(take).toBeCalledWith(actions.showAlertMessage.request(newAlert).type)
  //   generator.next([])
  //   expect(select).toBeCalledTimes(1)
  //   generator.next()
  //   expect(call).toBeCalledTimes(1)
  //   expect(call).toBeCalledWith(actions.showAlertMessage.success(newAlert))
  // })

  it('[Success] Should emit handleShowAlerts', () => {
    const newAlert = { ...SIMPLE_ALERT }
    const generator = handleShowAlerts(actions.showAlertMessage.request(newAlert).payload)
    generator.next()
    expect(put).toBeCalledTimes(1)
    expect(put).toBeCalledWith(actions.showAlertMessage.success(newAlert))
    expect(generator.next().done).toBe(true)
  })

  it('[Failed] Should emit handleShowAlerts', () => {
    const newAlert = { ...SIMPLE_ALERT }
    call.mockImplementation(() => {
      throw {
        code: 100,
        message: 'SOME_ERROR',
      }
    })
    put.mockImplementation(action => action)
    const generator = handleShowAlerts(actions.showAlertMessage.request(newAlert).payload)
    expect(generator.next()).toMatchSnapshot()
    expect(generator.next().done).toBe(true)
  })

  it('[Success] Should handle clearAlertMessage Request', () => {
    const generator = clearAlertsSaga()
    generator.next()
    expect(take).toBeCalledTimes(1)
    expect(take).toBeCalledWith(actions.clearAlertMessage.request().type)
    generator.next()
    expect(call).toBeCalledTimes(1)
  })
})
