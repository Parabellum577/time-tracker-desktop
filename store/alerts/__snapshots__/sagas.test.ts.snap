// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Store] Alert Sagas [Failed] Should emit handleClearAlert 1`] = `
Object {
  "done": false,
  "value": undefined,
}
`;

exports[`[Store] Alert Sagas [Failed] Should emit handleShowAlerts 1`] = `
Object {
  "done": false,
  "value": Object {
    "payload": Object {
      "alertMessage": "INVALID_LOGIN",
      "alertType": "error",
    },
    "type": "notifications/SHOW_ALERT_MESSAGE_SUCCESS",
  },
}
`;

exports[`[Store] Alert Sagas [Success] Should emit handleClearAlert 1`] = `
Object {
  "done": false,
  "value": undefined,
}
`;
